### PR TITLE
feat: Block readiness until caches are synced

### PR DESCRIPTION
**Pull request by ellistarn**
Fixes #N/A 

**Description**

Improve readiness checks to include cache hydration. Currently, readiness is blocked on webhooks, but which can become ready before the manager, giving us a false readiness signal.

**How was this change tested?**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -136,6 +136,11 @@ func NewOperator() (context.Context, *Operator) {
 		return []string{o.(*v1alpha5.Machine).Status.ProviderID}
 	}), "failed to setup machine provider id indexer")
 
+	lo.Must0(mgr.AddReadyzCheck("manager", func(req *http.Request) error {
+		mgr.GetCache().WaitForCacheSync(req.Context())
```
Comment (jonathan-innis) [inline pkg/operator/operator.go:140]: Does this hold the thread until the cache is synced?

Comment (jonathan-innis) [reply pkg/operator/operator.go:140]: If so, you probably want to do the check `IsCacheSynced` rather than `WaitForCacheSynced`

**Comment by ellistarn** [comment]
Did some more testing on this. Added a sleep on startup, to simulate a slow cache sync:
```
go func() {
		defer wg.Done()
		time.Sleep(60 * time.Second)
		lo.Must0(o.Manager.Start(ctx))
	}()
```

```
❯ k get pods -n karpenter -w
NAME                         READY   STATUS    RESTARTS   AGE
karpenter-5bc6cc5bc4-hdjlz   0/1     Pending   0          0s
karpenter-5bc6cc5bc4-hdjlz   0/1     Pending   0          0s
karpenter-5bc6cc5bc4-zgznv   0/1     Pending   0          0s
karpenter-5bc6cc5bc4-zgznv   0/1     Pending   0          0s
karpenter-5bc6cc5bc4-hdjlz   0/1     ContainerCreating   0          0s
karpenter-5bc6cc5bc4-zgznv   0/1     ContainerCreating   0          0s
karpenter-5bc6cc5bc4-zgznv   0/1     Running             0          3s
karpenter-5bc6cc5bc4-hdjlz   0/1     Running             0          3s
karpenter-5bc6cc5bc4-zgznv   1/1     Running             0          66s
karpenter-5bc6cc5bc4-hdjlz   1/1     Running             0          66s
```
```
k get events -n karpenter -w 
0s          Normal    Created                           pod/karpenter-5bc6cc5bc4-zgznv    Created container controller
1s          Normal    Created                           pod/karpenter-5bc6cc5bc4-hdjlz    Created container controller
0s          Normal    Started                           pod/karpenter-5bc6cc5bc4-zgznv    Started container controller
0s          Normal    Started                           pod/karpenter-5bc6cc5bc4-hdjlz    Started container controller
0s          Warning   Unhealthy                         pod/karpenter-5bc6cc5bc4-hdjlz    Readiness probe failed: Get "http://192.168.27.137:8081/readyz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
0s          Warning   Unhealthy                         pod/karpenter-5bc6cc5bc4-zgznv    Readiness probe failed: Get "http://192.168.66.182:8081/readyz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
0s          Normal    LeaderElection                    lease/karpenter-leader-election   karpenter-5bc6cc5bc4-zgznv_84e55191-7af2-46e8-93c4-5cd74736e696 became leader
```

File: pkg/operator/operator.go (hunk 1 above)
Comment (ellistarn) [reply pkg/operator/operator.go]: It does hang, but hanging is caught by the readiness probe's client timeout.

**Review by jonathan-innis — APPROVED** [review]
Nice! 🚀 LGTM!

**Pull request merged**